File: datasets/C++/Mistral/Algorithm/medium4.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

bool canPlaceWord(vector<string>& grid, const string& word, int row, int col, bool horizontal) {
    int length = word.length();
    if (horizontal) {
        if (col + length > 10) return false;
        for (int i = 0; i < length; ++i) {
            if (grid[row][col + i] != '-' && grid[row][col + i] != word[i]) {
                return false;
            }
        }
    } else {
        if (row + length > 10) return false;
        for (int i = 0; i < length; ++i) {
            if (grid[row + i][col] != '-' && grid[row + i][col] != word[i]) {
                return false;
            }
        }
    }
    return true;
}

void placeWord(vector<string>& grid, const string& word, int row, int col, bool horizontal) {
    int length = word.length();
    if (horizontal) {
        for (int i = 0; i < length; ++i) {
            grid[row][col + i] = word[i];
        }
    } else {
        for (int i = 0; i < length; ++i) {
            grid[row + i][col] = word[i];
        }
    }
}

void unplaceWord(vector<string>& grid, const string& word, int row, int col, bool horizontal) {
    int length = word.length();
    if (horizontal) {
        for (int i = 0; i < length; ++i) {
            grid[row][col + i] = '-';
        }
    } else {
        for (int i = 0; i < length; ++i) {
            grid[row + i][col] = '-';
        }
    }
}

bool solveCrossword(vector<string>& grid, const vector<string>& words, int index) {
    if (index == words.size()) return true;

    const string& word = words[index];
    for (int row = 0; row < 10; ++row) {
        for (int col = 0; col < 10; ++col) {
            if (grid[row][col] == '-' || grid[row][col] == word[0]) {
                if (canPlaceWord(grid, word, row, col, true)) {
                    placeWord(grid, word, row, col, true);
                    if (solveCrossword(grid, words, index + 1)) return true;
                    unplaceWord(grid, word, row, col, true);
                }
                if (canPlaceWord(grid, word, row, col, false)) {
                    placeWord(grid, word, row, col, false);
                    if (solveCrossword(grid, words, index + 1)) return true;
                    unplaceWord(grid, word, row, col, false);
                }
            }
        }
    }
    return false;
}

vector<string> crosswordPuzzle(vector<string>& crossword, const string& wordsInput) {
    vector<string> words;
    string word;
    for (char c : wordsInput) {
        if (c == ';') {
            words.push_back(word);
            word.clear();
        } else {
            word += c;
        }
    }
    if (!word.empty()) words.push_back(word);

    sort(words.begin(), words.end(), [](const string& a, const string& b) {
        return a.length() > b.length();
    });

    if (solveCrossword(crossword, words, 0)) {
        return crossword;
    } else {
        return {"No solution found"};
    }
}

int main() {
    vector<string> crossword(10);
    cout << "Enter the crossword grid (10 rows of 10 characters each, use '-' for fillable cells and '+' for blocked cells):" << endl;
    for (int i = 0; i < 10; ++i) {
        while (true) {
            cout << "Row " << i + 1 << ": ";
            cin >> crossword[i];
            if (crossword[i].length() == 10 && all_of(crossword[i].begin(), crossword[i].end(), [](char c) { return c == '+' || c == '-'; })) {
                break;
            } else {
                cout << "Invalid input. Please enter exactly 10 characters, each being either '+' or '-'." << endl;
            }
        }
    }

    string words;
    cout << "Enter the list of words separated by semicolons: ";
    cin >> words;

    vector<string> solution = crosswordPuzzle(crossword, words);
    cout << "\nSolution:" << endl;
    for (const string& row : solution) {
        cout << row << endl;
    }

    return 0;
}

File: datasets/Python/Mistral/Algorithm/medium4.py
def can_place_word(grid, word, row, col, direction):
    length = len(word)
    if direction == 'H':  # Horizontal
        if col + length > 10:
            return False
        for i in range(length):
            if grid[row][col + i] not in ('-', word[i]):
                return False
    elif direction == 'V':  # Vertical
        if row + length > 10:
            return False
        for i in range(length):
            if grid[row + i][col] not in ('-', word[i]):
                return False
    return True

def place_word(grid, word, row, col, direction):
    length = len(word)
    if direction == 'H':  # Horizontal
        for i in range(length):
            grid[row][col + i] = word[i]
    elif direction == 'V':  # Vertical
        for i in range(length):
            grid[row + i][col] = word[i]

def unplace_word(grid, word, row, col, direction):
    length = len(word)
    if direction == 'H':  # Horizontal
        for i in range(length):
            grid[row][col + i] = '-'
    elif direction == 'V':  # Vertical
        for i in range(length):
            grid[row + i][col] = '-'

def solve_crossword(grid, words, index):
    if index == len(words):
        return True

    word = words[index]
    for row in range(10):
        for col in range(10):
            if grid[row][col] == '-' or grid[row][col] == word[0]:
                # Try placing horizontally
                if can_place_word(grid, word, row, col, 'H'):
                    place_word(grid, word, row, col, 'H')
                    if solve_crossword(grid, words, index + 1):
                        return True
                    unplace_word(grid, word, row, col, 'H')

                # Try placing vertically
                if can_place_word(grid, word, row, col, 'V'):
                    place_word(grid, word, row, col, 'V')
                    if solve_crossword(grid, words, index + 1):
                        return True
                    unplace_word(grid, word, row, col, 'V')
    return False

def crosswordPuzzle(crossword, words):
    # Sort words by length in descending order
    words = sorted(words.split(';'), key=len, reverse=True)

    # Convert crossword grid to a list of lists for easier manipulation
    grid = [list(row) for row in crossword]

    # Solve the crossword
    if solve_crossword(grid, words, 0):
        # Convert the grid back to strings
        return [''.join(row) for row in grid]
    else:
        return ["No solution found"]

def main():
    # Read the crossword grid
    crossword = []
    print("Enter the crossword grid (10 rows of 10 characters each, use '-' for fillable cells and '+' for blocked cells):")
    for i in range(10):
        while True:
            row = input(f"Row {i + 1}: ").strip()
            if len(row) == 10 and all(c in '+-' for c in row):
                crossword.append(row)
                break
            else:
                print("Invalid input. Please enter exactly 10 characters, each being either '+' or '-'.")

    # Read the list of words
    words = input("Enter the list of words separated by semicolons: ").strip()

    # Solve the crossword puzzle
    solution = crosswordPuzzle(crossword, words)
    print("\nSolution:")
    for row in solution:
        print(row)

if __name__ == "__main__":
    main()
